### PR TITLE
Port CompareTo and GetTypeCode for primitive types

### PR DESCRIPTION
**Pull request by danmosemsft**
Completed @Lixire 's work from https://github.com/dotnet/corefx/pull/11134 for #10137

@ellismg @joperezr

### CODE REVIEW

File: src/System.Runtime/tests/GetTypeCodeTests.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        }
+
+        [Fact]
+        public static void FromUInt64()
+        {
+            var val = Int64.MinValue;
+            TypeCode testValue = val.GetTypeCode();
+            Assert.Equal(TypeCode.Int64, testValue);
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: There's a lot of duplication in these types.  A theory would help, e.g. something like
```C#
[Theory]
[InlineData(TypeCode.Boolean, false)]
[InlineData(TypeCode.Char, 'c')]
[InlineData(TypeCode.Double, 42.0)]
...
public static void GetTypeCode_Primitives(TypeCode expected, IConvertible convertible)
{
    Assert.Equal(expected, convertible.GetTypeCode());
}
```

Comment (danmosemsft) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: Good point, I did this for all the types that C# accepts suffixes for (a cast is not the required constant expression) excepting Decimal. For some reason it doesn't consider 42M to be a constant decimal expressoin, and I can't figure out why from the language spec.

---
In reply to: [76539432](https://github.com/dotnet/corefx/pull/11210#discussion_r76539432) [](ancestors = 76539432)

Comment (stephentoub) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: > For some reason it doesn't consider 42M to be a constant decimal expressoin

decimal is not a CLR primitive type and as such can't be used in an attribute, unfortunately

Comment (justinvp) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: MemberData instead InlineData?

```c#
public static object[][] TypeCodeTestData =
{
    new object[] { TypeCode.Boolean, false },
    new object[] { TypeCode.Char, 'x' },
    new object[] { TypeCode.Double, 42.0 },
    new object[] { TypeCode.Int32, 42 },
    new object[] { TypeCode.Int64, 42L },
    new object[] { TypeCode.Single, 42f },
    new object[] { TypeCode.UInt32, 42u },
    new object[] { TypeCode.UInt64, 42ul },
    new object[] { TypeCode.Byte, (byte)42 },
    new object[] { TypeCode.Decimal, 42m },
    // TODO add the rest
};

[Theory]
[MemberData(nameof(TypeCodeTestData))]
public static void GetTypeCode_Primitives(TypeCode expected, IConvertible convertible)
{
    Assert.Equal(expected, convertible.GetTypeCode());
}
```

File: src/System.Runtime/tests/GetTypeCodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        public static void FromSByte()
+        {
+            Assert.Equal(TypeCode.SByte, sbyte.MinValue.GetTypeCode());
+        }
+
+        [Fact]
+        public static void FromUShort()
+        {
+            Assert.Equal(TypeCode.UInt16, ushort.MinValue.GetTypeCode());
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: Why did you keep all of these separate?  I understand why for decimal, but the rest should all be encodable in the attribute, e.g.
```[InlineData(TypeCode.Byte, (byte)42)]```

**Comment by stephentoub** [comment]
Other than that one nit, LGTM

File: src/System.Runtime/tests/GetTypeCodeTests.cs (hunk 2 above)
Comment (danmosemsft) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: I mistakenly concluded that the cast was prohibited, and those don't have suffixes. Fixed.

Can you explain why Decimal is special? I can't find any suggestion why, in the C# or ECMA specifications.

Comment (stephentoub) [inline src/System.Runtime/tests/GetTypeCodeTests.cs]: > Can you explain why Decimal is special?

It's not... that's the point ;)

There is nothing special about Decimal as far as the runtime is concerned.  Primitives like Byte, Char, Int32, etc. are special in that they can be encoded in MSIL metadata.  That's not the case for arbitrary value types, of which Decimal is one as far as the runtime is concerned.  It so happens that the C# language chose to special case Decimal by giving it a language keyword (```decimal```) and a modifier for constructing values (```m```), but it's only special as far as the language is concerned.

**Comment by joperezr** [comment]
LGTM assuming the tests ran successful locally.

**Comment by danmosemsft** [comment]
'''08:42:03   Unhandled Exception: System.IO.FileLoadException: Could not load file or assembly 'System.IO, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)'''

This looks like a race in the build.

@dotnet-bot test Innerloop CentOS7.1 Release Build and Test

**Comment by stephentoub** [comment]
@danmosemsft, yes, dev/api is hitting this a lot... I opened https://github.com/dotnet/corefx/issues/11203 for it over the weekend.

**Pull request merged**